### ISSUE TITLE
Using Generic and __new__ in class def, mypy misses incorrect parameter if parameter created inline

### ISSUE THREAD

**Issue by Dr-Irv** [python/mypy#14009]
**Bug Report**

This came up in pandas-stubs, issue https://github.com/pandas-dev/pandas-stubs/issues/412

If you have a stub that defines a `Generic` class with `__new__()`, with overloads, and the final overload doesn't specify the type of the generic, then a function call that creates an object of the generic class in the function call is not picked up as an error, but it is picked up if stored in another variable.

**To Reproduce**

Two files are needed.  First is `gptst2.pyi`  :
```python
from typing import Generic, TypeVar, Type, overload

TV = TypeVar("TV", int, str)

class Foo(Generic[TV]):
    @overload
    def __new__(cls, ty: Type[TV]) -> Foo[TV]: ...
    @overload
    def __new__(cls, ty=...) -> Foo: ...
```
Then the tester code `genparam.py`:
```python
from gptst2 import Foo


def fun(param: Foo[int]) -> Foo[str]:
    ...


fun(param=Foo(str))
p = Foo(str)
fun(p)
```

**Expected Behavior**

mypy should report errors on both calls to `fun()`

**Actual Behavior**
mypy does not pick up that the first call to `fun` is incorrect.  It does pick up that the second call is incorrect.
pyright picks up both as being incorrect.

Output from mypy:
```text
genparam.py:10: error: Argument 1 to "fun" has incompatible type "Foo[str]"; expected "Foo[int]"
```

**Your Environment**

- Mypy version used:  0.971
- Mypy command-line flags:  `--no-incremental`
- Mypy configuration options from `mypy.ini` (and other config files):  None
- Python version used:  3.10

**Comment by erictraut** [comment]
I think mypy is correct here. The behavior you cited for pyright was due to a bug which was fixed a while ago. Pyright and mypy now agree.

**Comment by JelleZijlstra** [comment]
I believe it's due to type context (Eric calls the equivalent technique in pyright 'bidirectional inference'). When a particular type is expected, type checkers try to infer that type to minimize errors. However, when the value is assigned to a variable, that assignment is type checked independently and sets the type.

Consider code like this:

```python
def f(x: list[object]) -> None: pass

y = [1]
f(y)

f([1])
```

Here mypy shows an error for the first call, but not the second.

**Comment by Dr-Irv** [comment]
> When a particular type is expected, type checkers try to infer that type to minimize errors. However, when the value is assigned to a variable, that assignment is type checked independently and sets the type.

I don't think your example fits in this case, because `object` is more of a supertype.

In my example, you have
```python
def fun(param: Foo[int]) -> Foo[str]:
    ...

fun(param=Foo(str))
p = Foo(str)
fun(p)
```

So I don't see how `Foo(str)`, which has type `Foo[str]` could ever match `Foo[int]` in the first call to `fun()`

**Comment by erictraut** [comment]
When you call `fun(param=Foo(str))` (or the equivalent `fun(Foo(str))`), a type checker needs to determine whether there a way for the expression `Foo(str)` to be evaluated such that it can satisfy the call to `fun`. In this case, there is a way because there is an overload that returns a `Foo[Any]`, and `Foo[Any]` is compatible with `Foo[int]`.

By contrast, if you construct a `Foo[str]` independent of the call to `fun`, there is no way to transform that `Fun[str]` into a `Fun[int]` such that it's compatible with parameter `param`.

It's similar to this:
```python
a1 = [1, 2, 3]
x1: list[float] = a # Type error

x2: list[float] = [1, 2, 3] # No type error
```
